### ISSUE TITLE
ios safari webp show error

### ISSUE THREAD

**Issue by sunxiang520** [webp-sh/webp_server_go#61]
> ios device display webp error
device:
iphone se   ios14.2

webp_server_go version:0.3.0
```
request:
Accept: image/webp,image/png,image/svg+xml,image/*;q=0.8,video/*;q=0.8,*/*;q=0.5
User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1
Accept-Language: zh-cn
Accept-Encoding: gzip, deflate
Connection: keep-alive

response:
Response Code  404 Not Found
```

**Comment by BennyThink** [comment]
Please check whether your original image(png, jpg) exists.

**Comment by sunxiang520** [comment]
1. original image  exists,I use AWS S3  Remote Backend server,only ios device show error,android device show success.

2. when I use local image path,ios device show success,but server response image/jpg, due to ios14 support webp, server should  response image/webp.

**Comment by BennyThink** [comment]
Okay, the first one seems weird. Maybe @n0vad3v and I will take a look on weekends.

As to the second one, the truth is, we currently do not have an effective mechanism to identify where the user is iOS 14 or prior to it. So we choose to send original image for all iOS/iPadOS devices. That would be awesome if you could come up with a realistic identify mechanism.

**Comment by n0vad3v** [comment]
Hello @sunxiang520 , for the first part, could you help provide a sample S3 address with the same configuration related to the problem and provide the `config.json` you are using?

**Comment by chinaapus** [comment]
@BennyThink 

Accept: image/webp,image/png,image/svg+xml,image/*;q=0.8,video/*;q=0.8,*/*;q=0.5

`Accept: image/webp`


Modern browsers will send an Accept indicating webp, it is absolutely possible to recognize this

**Comment by BennyThink** [comment]
Wow, excellent idea. We'll implement in the next version

**Comment by BennyThink** [comment]
@chinaapus 
Webkit seems weird to me.

# macOS Safari
```
GET /webp_server.jpg HTTP/1.1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Safari/605.1.15
Host: 192.168.7.67:3333
Content-Length: 0
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: zh-cn
Accept-Encoding: gzip, deflate
Connection: keep-alive
```

# iPhone iOS 14 Safari
```
GET /webp_server.jpg HTTP/1.1
User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1
Host: 192.168.7.67:3333
Content-Length: 0
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: zh-cn
Accept-Encoding: gzip, deflate
Connection: keep-alive
```

# iPhone iOS 14 Chrome
```
# GET /webp_server.jpg HTTP/1.1
User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/87.0.4280.77 Mobile/15E148 Safari/604.1
Host: 192.168.7.67:3333
Content-Length: 0
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: zh-cn
Accept-Encoding: gzip, deflate
Connection: keep-alive
```
# iPad OS 14 Safari
```
GET /webp_server.jpg HTTP/1.1
User-Agent: Mozilla/5.0 (iPad; CPU OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1
Host: 192.168.7.67:3333
Content-Length: 0
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: zh-cn
Accept-Encoding: gzip, deflate
Connection: keep-alive
```
# iPad OS 14 Chrome
```
GET /webp_server.jpg HTTP/1.1
User-Agent: Mozilla/5.0 (iPad; CPU OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/87.0.4280.77 Mobile/15E148 Safari/604.1
Host: 192.168.7.67:3333
Content-Length: 0
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: zh-cn
Accept-Encoding: gzip, deflate
Connection: keep-alive
```

-------
They don't even send effective `accept` headers. Webkit is really weird. I think it's not so absolute possible to detect this - maybe I'll have to start with UA.

**Comment by sunxiang520** [comment]
> @chinaapus
> Webkit seems weird to me.
> 
> # macOS Safari
> ```
> GET /webp_server.jpg HTTP/1.1
> User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Safari/605.1.15
> Host: 192.168.7.67:3333
> Content-Length: 0
> Upgrade-Insecure-Requests: 1
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
> Accept-Language: zh-cn
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> ```
> 
> # iPhone iOS 14 Safari
> ```
> GET /webp_server.jpg HTTP/1.1
> User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1
> Host: 192.168.7.67:3333
> Content-Length: 0
> Upgrade-Insecure-Requests: 1
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
> Accept-Language: zh-cn
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> ```
> 
> # iPhone iOS 14 Chrome
> ```
> # GET /webp_server.jpg HTTP/1.1
> User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/87.0.4280.77 Mobile/15E148 Safari/604.1
> Host: 192.168.7.67:3333
> Content-Length: 0
> Upgrade-Insecure-Requests: 1
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
> Accept-Language: zh-cn
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> ```
> 
> # iPad OS 14 Safari
> ```
> GET /webp_server.jpg HTTP/1.1
> User-Agent: Mozilla/5.0 (iPad; CPU OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1
> Host: 192.168.7.67:3333
> Content-Length: 0
> Upgrade-Insecure-Requests: 1
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
> Accept-Language: zh-cn
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> ```
> 
> # iPad OS 14 Chrome
> ```
> GET /webp_server.jpg HTTP/1.1
> User-Agent: Mozilla/5.0 (iPad; CPU OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/87.0.4280.77 Mobile/15E148 Safari/604.1
> Host: 192.168.7.67:3333
> Content-Length: 0
> Upgrade-Insecure-Requests: 1
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
> Accept-Language: zh-cn
> Accept-Encoding: gzip, deflate
> Connection: keep-alive
> ```
> 
> macOS Chrome
> 
> ```
> GET /webp_server.jpg HTTP/1.1
> User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.88 Safari/537.36
> Host: 192.168.7.67:3333
> Content-Length: 0
> Connection: keep-alive
> Cache-Control: max-age=0
> Dnt: 1
> Upgrade-Insecure-Requests: 1
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
> Accept-Encoding: gzip, deflate
> Accept-Language: zh-CN,zh;q=0.9,en-US;q=0.8,en;q=0.7,zh-TW;q=0.6
> If-None-Match: W/"6638-23BD942B"
> If-Modified-Since: Fri, 27 Nov 2020 06:26:04 GMT
> ```
> 
> They don't even send effective `accept` headers. Webkit is really weird. I think it's not so absolute possible to detect this - maybe I'll have to start with UA.

Because you directly access the image url.accept` headers not show image/webp,Loading the image url in the website will display image/webp.You can have a try.

**Comment by BennyThink** [comment]
Oooops, it seems that way.

**Comment by BennyThink** [comment]
@sunxiang520 the second part of your issue has been resolved on master branch.
We'll try to track the first part if you could generously provide the necessities

* sample S3 address with the same configuration related to the problem 
* provide the config.json

**Comment by sunxiang520** [comment]
> I use minio build object storage server,only ios device show error,android device show success.

test image url:
https://pic.youku-cdn1.xyz/test/202008211597973938.jpg
oss original image url:
https://oss.youku-cdn1.xyz/test/202008211597973938.jpg
config.json
```
{
    "HOST": "127.0.0.1",
    "PORT": "3333",
    "QUALITY": "75",
    "IMG_PATH": "https://oss.youku-cdn1.xyz",
    "EXHAUST_PATH": "/www/wwwroot/webp/cache",
    "ALLOWED_TYPES": ["jpg","png","jpeg","gif"]
}
````

> @sunxiang520 the second part of your issue has been resolved on master branch.
> We'll try to track the first part if you could generously provide the necessities
> 
> * sample S3 address with the same configuration related to the problem
> * provide the config.json



> @sunxiang520 the second part of your issue has been resolved on master branch.
> We'll try to track the first part if you could generously provide the necessities
> 
> * sample S3 address with the same configuration related to the problem
> * provide the config.json

**Comment by sunxiang520** [comment]
> @sunxiang520 the second part of your issue has been resolved on master branch.
> We'll try to track the first part if you could generously provide the necessities
> 
> * sample S3 address with the same configuration related to the problem
> * provide the config.json

> I use minio build object storage server,only ios device show error,android device show success.

test image url:
https://pic.youku-cdn1.xyz/test/202008211597973938.jpg
oss original image url:
https://oss.youku-cdn1.xyz/test/202008211597973938.jpg
config.json
```
{
    "HOST": "127.0.0.1",
    "PORT": "3333",
    "QUALITY": "75",
    "IMG_PATH": "https://oss.youku-cdn1.xyz",
    "EXHAUST_PATH": "/www/wwwroot/webp/cache",
    "ALLOWED_TYPES": ["jpg","png","jpeg","gif"]
}
````

**Comment by BennyThink** [comment]
@sunxiang520 confirm bug, will fix in #64, waiting for review from @n0vad3v
